services/config.go

```diff
@@ -0,0 +1,264 @@
+	}
+
+	if opts.PollInterval <= 0 {
+		opts.PollInterval = defaultConfigPollInterval
+	}
+
+	if opts.PollJitter <= 0 {
+		opts.PollJitter = defaultConfigPollJitter
+	}
+
+	logger := golog.LoggerFor("configservice")
+
+	u, err := url.Parse(opts.OriginURL)
+	if err != nil {
+		logger.Fatalf("Unable to parse chained cloud config URL: %v", err)
+	}
+
+	detour.ForceWhitelist(u.Host)
+
+	userId := strconv.Itoa(int(opts.UserConfig.GetUserID()))
+	_configService.opts = opts
+	_configService.clientInfo = &apipb.ConfigRequest_ClientInfo{
+		FlashlightVersion: common.LibraryVersion,
+		ClientVersion:     common.CompileTimeApplicationVersion,
+		UserId:            userId,
+		ProToken:          opts.UserConfig.GetToken(),
+	}
+	_configService.done = make(chan struct{})
+	_configService.logger = logger
+
+	config := handler.GetConfig()
+	_configService.clientInfo.Country = config.GetCountry()
+
+	_configService.logger.Debug("Starting config service")
+	_configService.running = true
+	if opts.Sticky {
+		return _configService.Stop, nil
+	}
+
+	fn := func() int64 {
+		sleep, _ := _configService.fetchConfig()
+		return sleep
+	}
+	go callRandomly(fn, opts.PollInterval, opts.PollJitter, _configService.done, _configService.logger)
+
+	return _configService.Stop, nil
+}
+
+func (cs *configService) updateClientInfo(conf *apipb.ConfigResponse) {
+	cs.clientInfo.ProToken = conf.ProToken
+	cs.clientInfo.Country = conf.Country
+	cs.clientInfo.Ip = conf.Ip
+}
+
+func (cs *configService) Stop() {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if !cs.running {
+		return
+	}
+
+	close(cs.done)
+	cs.running = false
+}
+
+// fetchConfig fetches the current config from the server and updates the client's config if a change
+// has occurred. It returns the extra sleep time received from the server response and any error that
+// occurred.
+func (cs *configService) fetchConfig() (int64, error) {
+	op := ops.Begin("Fetching config")
+	defer op.End()
+
+	newConf, sleep, err := cs.fetch()
+	if err != nil {
+		return 0, op.FailIf(err)
+	}
+
+	cs.lastFetched = time.Now()
+
+	cs.logger.Debug("Received config")
+	curConf := cs.configHandler.GetConfig()
+	if curConf != nil && !configIsNew(curConf, newConf) {
+		op.Set("config_changed", false)
+		cs.logger.Debug("Config is unchanged")
+		return sleep, nil
+	}
+
+	op.Set("config_changed", true)
+
+	cs.updateClientInfo(newConf)
+	cs.configHandler.SetConfig(newConf)
+
+	return sleep, nil
+}
+
+func (cs *configService) fetch() (*apipb.ConfigResponse, int64, error) {
+	confReq := cs.newRequest()
+	buf, err := proto.Marshal(confReq)
+	if err != nil {
+		return nil, 0, fmt.Errorf("unable to marshal config request: %w", err)
+	}
+
+	resp, sleep, err := post(
+		cs.opts.OriginURL,
+		bytes.NewReader(buf),
+		cs.opts.RoundTripper,
+		cs.opts.UserConfig,
+		cs.logger,
+	)
+	if err != nil {
+		return nil, 0, fmt.Errorf("config request failed: %w", err)
+	}
+	defer resp.Close()
+
+	gzReader, err := gzip.NewReader(resp)
```

Comment:
Are protocol buffer responses from the back-end gzipped?  That'd be a bit odd I think, but I guess I don't actually know.

Reply:
Why would that be odd? I would assume everything (except maybe single TCP request/response) would be gzipped since it reduces bandwidth. I don't actually know what the text form of a protocol buffer looks like, so maybe it's already very small.

Reply:
The gzipping was from when responses were returned as json/yaml, but I'm unclear if gzipping makes proto smaller or not.

Reply:
Alright. Well, I'll just remove it for now. We can always add it back in later if we want.